marlowe-playground-client/src/Marlowe/Semantics.purs

```diff
@@ -1210,26 +1212,41 @@ moneyInContract state =
     (\(Tuple _ (Token cur tok)) balance -> asset cur tok balance)
     (unwrap state).accounts
 
-class HasMaxTime a where
-  maxTime :: a -> Timeout
+class HasTimeout a where
+  timeouts :: a -> { maxTime :: Timeout, minTime :: Maybe Timeout }
 
-instance hasMaxTimeContract :: HasMaxTime Contract where
-  maxTime Close = zero
-  maxTime (Pay _ _ _ _ contract) = maxTime contract
-  maxTime (If _ contractTrue contractFalse) = maxOf [ maxTime contractTrue, maxTime contractFalse ]
-  maxTime (When cases timeout contract) = maxOf [ maxTime cases, timeout, maxTime contract ]
-  maxTime (Let _ _ contract) = maxTime contract
-  maxTime (Assert _ contract) = maxTime contract
+instance hasTimeoutContract :: HasTimeout Contract where
+  timeouts Close = { maxTime: zero, minTime: Nothing }
+  timeouts (Pay _ _ _ _ contract) = timeouts contract
+  timeouts (If _ contractTrue contractFalse) =
+    let
+      ts = [ timeouts contractTrue, timeouts contractFalse ]
+    in
+      { maxTime: maxOf (map _.maxTime ts), minTime: minOf (map _.minTime ts) }
+  timeouts (When cases timeout contract) =
+    let
+      ts = [ timeouts cases, { maxTime: timeout, minTime: Just timeout }, timeouts contract ]
+    in
+      { maxTime: maxOf (map _.maxTime ts), minTime: minOf (map _.minTime ts) }
+  timeouts (Let _ _ contract) = timeouts contract
+  timeouts (Assert _ contract) = timeouts contract
 
-instance hasMaxTimeCase :: HasMaxTime Case where
-  maxTime (Case _ contract) = maxTime contract
+instance hasTimeoutCase :: HasTimeout Case where
+  timeouts (Case _ contract) = timeouts contract
 
-instance hasMaxTimeArray :: HasMaxTime a => HasMaxTime (Array a) where
-  maxTime = maxOf <<< map maxTime
+instance hasTimeoutArray :: HasTimeout a => HasTimeout (Array a) where
+  timeouts vs =
+    let
+      ts = map timeouts vs
+    in
+      { maxTime: maxOf (map _.maxTime ts), minTime: minOf (map _.minTime ts) }
 
 maxOf :: Array Timeout -> Timeout
 maxOf = foldl max zero
 
+minOf :: Array (Maybe Timeout) -> Maybe Timeout
+minOf as = minimum $ catMaybes as
+
```

Comment:
Just want to check, are these the right semantics?

If I have three jobs, with minimum timeouts of 1s, 5s & 10s, then if I combine them into a single job, what's the behaviour? Surely we're saying the combined job _must_ run for at least 10 seconds?

Not sure if the same logic applies here, but as I say, I want to check.

Reply:
here we want the result to be 1s. 
```haskell
minimum $ catMaybes [ Nothing, Just 1, Nothing, Just 5, Just 10 ]
> minimum [ 1, 5, 10 ]
> 1
```

Reply:
because we want to find the timeout that's coming next.

Reply:
Fair enough. 👍